RecipeApp.Core/RecipeConverter.cs

```diff
@@ -1,14 +1,176 @@
+            bool sorted = true;
+            for (int j = 0; j < multipliers.Length; j++)
+            {
+                for (int i = 0; i < multipliers.Length - 1; i++)
+                {
+                    if (multipliers[i]>multipliers[i+1])
+                    {
+                        //swapping multipliers
+                        double temp = multipliers[i];
+                        multipliers[i] = multipliers[i + 1];
+                        multipliers[i + 1] = temp;
+                        //swapping units
+                        string strTemp = units[i];
+                        units[i] = units[i + 1];
+                        units[i + 1] = strTemp;
+
+                        sorted = false;
+                    }
+                }
+                if (sorted)
+                {
+                    return;
+                }
+            }
+        }
+
+        // Validation method to check if string contains any of the cooking units
+        public static bool DoesContainCookingUnits(string textToCheck) {
+            bool containsUnits = false;
+            List<string> wordList = new List<string>(textToCheck.ToLower().Split(' ', '\n'));
+            for (int i = 0; i < units.Length; i++)
+            {
+                if (wordList.Contains(units[i]))
+                {
+                    containsUnits = true;
+                }
+            }
+            if (wordList.Contains("ml") || wordList.Contains("l"))
+            {
+                containsUnits = true;
+            }
+            return containsUnits;
+        }
+
+        // Check method to ensure that each unit as some value number before it
+        // returns -1 if ok, else returns index of bad cooking unit
+        public static int ReturnFirstFaultyCookingUnitIndex(string textToCheck) {
+            List<string> words = new List<string>(textToCheck.ToLower().Split(' ', '\n'));
+            textToCheck = FixDoubleWordedCookingUnits(textToCheck);
+            
+            // looking for cooking units
+            for (int i = 0; i < units.Length; i++)
+            {
+                string lookFor = units[i].Replace(" ", "_").ToLower();
+                int foundUnitIndex = words.FindIndex(x => x.Contains(lookFor));
+                if (foundUnitIndex != -1)
+                {
+                    if (foundUnitIndex != 0)
+                    {
+                        if (!ParseDouble(words[foundUnitIndex - 1], out var digitAmount))
+                        {
+                            return foundUnitIndex;
+                        }
+                        else {
+                            words.RemoveRange(foundUnitIndex - 1, 2);
+                            i--;
+                        }
+                    }
+                    else
+                    {
+                        return 0;
+                    }
+                }
+                else
+                {
+                    continue;
+                }
+            }
```

Comment:
Refactor:
```cs
for (int i = 0; i < units.Length; i++)
{
    string lookFor = units[i].Replace(" ", "_").ToLower();
    int foundUnitIndex = words.FindIndex(x => x.Contains(lookFor));

    if (foundUnitIndex == -1) continue;
    if (foundUnitIndex == 0) return 0;

    var isNumber = ParseDouble(words[foundUnitIndex - 1], out _);
    if (!isNumber) return foundUnitIndex;

    words.RemoveRange(foundUnitIndex - 1, 2);
    i--;
}
```
Note: `_` is a discard operator- used when you don't care about the out- you just want the normal return, like in your case- just to see if Tryparse returns true or false. More info: https://docs.microsoft.com/en-us/dotnet/csharp/discards